nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/GetAWSIoTShadow.java

```diff
@@ -0,0 +1,114 @@
+        "when overriding the processor-configuration with a message-attribute \"aws.iot.thing.override\".")
+@SeeAlso({ ConsumeAWSIoTMqtt.class })
+@ReadsAttributes({
+        @ReadsAttribute(attribute = "aws.iot.thing.override", description = "Overrides the processor configuration for topic."),
+})
+@WritesAttributes({
+        @WritesAttribute(attribute = "aws.iot.thing", description = "Thing name in AWS IoT"),
+})
+public class GetAWSIoTShadow extends AbstractAWSIoTShadowProcessor {
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(
+                    PROP_THING,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE,
+                    TIMEOUT,
+                    PROXY_HOST,
+                    PROXY_HOST_PORT,
+                    REGION));
+
+    private final static String ATTR_NAME_THING = PROP_NAME_THING + ".override";
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return Collections.singleton(REL_SUCCESS);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final AWSIotDataClient iotClient = this.getClient();
+
+        if (iotClient == null) {
+            getLogger().error("AWS-Client was not initialized. See logs to find reasons.");
+            return;
+        }
+        // get flowfile
+        FlowFile flowFile = session.get();
+        // if provided override configured thing name with name from corresponding message attribute
+        String thingName = flowFile != null && flowFile.getAttributes().containsKey(ATTR_NAME_THING)
+                        ? flowFile.getAttribute(ATTR_NAME_THING)
+                        : context.getProperty(PROP_NAME_THING).getValue();
+
+        FlowFile flowFileOut = flowFile == null ? session.create() : flowFile;
+
+        // ask shadow of the thing for last reported state by requesting the API of AWS
+        final GetThingShadowRequest iotRequest = new GetThingShadowRequest().withThingName(thingName);
+        final GetThingShadowResult iotResponse = iotClient.getThingShadow(iotRequest);
+
+        //FlowFile flowFileOut = session.create();
+        final Map<String, String> attributes = new HashMap<>();
+        attributes.put(PROP_NAME_THING, thingName);
+        flowFileOut = session.putAllAttributes(flowFileOut, attributes);
+
+        flowFileOut = session.write(flowFileOut, new OutputStreamCallback() {
+            @Override
+            public void process(final OutputStream out) throws IOException {
+                out.write(iotResponse.getPayload().array());
+            }
+        });
+        session.transfer(flowFileOut, REL_SUCCESS);
+        session.commit();
```

Comment:
**JPercivall**
Should emit a "FETCH" provenance event when created using an incoming flowfile and "RECEIVE" when the processor is acting as a source.